### PR TITLE
Backport fix for MethodImpl removal when interfaceImpl is removed

### PR DESCRIPTION
**Pull request by jtschuster**
Backports https://github.com/dotnet/runtime/pull/102857 to .NET 8.

### Customer Impact:

The issue was first reported in https://github.com/dotnet/runtime/issues/102796. This affects situations where a static interface method definition is kept, an implementing type is called only directly on the declaring type (`TypeName.Method()` is called rather than `T.Method()` `where T : IFoo`), and the InterfaceImpl is removed. The MethodImpl pointing to the interface method is kept, but since the type doesn't implement the interface anymore, a TypeLoadException is thrown at runtime.

There is a workaround that requires creating an explicit interface implementation method that forwards to the public method (see https://github.com/dotnet/runtime/issues/102796#issuecomment-2164009664). It's not a bad workaround, but finding the workaround wouldn't be easy.

### Testing:

A new test is added which exercises this scenario and checks for the proper removal of the MethodImpl. Assertions related to MethodImpl removal in other tests have been reenabled.

### Risk:

There is a small chance that this change introduces a new bug. Interface and override trimming have been the primary source of issues reported in the trimmer. However, this change is fairly small and hasn't caused issues in main. It seems worth the risk to fix this scenario.

### CODE REVIEW

**Review by sbomer — APPROVED** [review]
Code change LGTM.

To me this feels riskier than the average backport we've taken for illink, since it touches an area that has been a source of bugs (as you mentioned) - and especially because that area has evolved in .NET 9 and we're not getting as much test coverage of the old behavior in the net8 branch. Personally I would wait to see if there's another customer report before we take this.

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/StaticInterfaceMethods/RemovedInterfaceImplementationRemovedOverride.cs [hunk 1]
```diff
@@ -22,6 +22,8 @@ public static void Main ()
 			GenericInterfaceGenericType.Test ();
 			PrivateExplicitImplementationReflectedOver.Test ();
 			InheritedInterfaces.Test ();
+			DimWithOverride.Test ();
+			Generics.Test ();
```
[the remark this answers is not in the document]
Comment (jtschuster) [reply src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/StaticInterfaceMethods/RemovedInterfaceImplementationRemovedOverride.cs:26]: These changes are in https://github.com/dotnet/runtime/pull/103317/files#diff-0e98f4278be59be3dd229120370b46a5ccaaec54d9c63ad6cdf091a594adbaa1

**Comment by jtschuster** [comment]
> To me this feels riskier than the average backport we've taken for illink, since it touches an area that has been a source of bugs (as you mentioned) - and especially because that area has evolved in .NET 9 and we're not getting as much test coverage of the old behavior in the net8 branch. Personally I would wait to see if there's another customer report before we take this.

I agree, I'd be okay leaving this as a draft for now.

**Comment by agocke** [comment]
Yeah, I'd probably hold off for now. Alternatively, we could try to find a bigger test suite that we could run against that we think would have a reasonable chance of revealing the bug.

**Comment by carlossanlop** [comment]
@jtschuster friendly reminder that code complete for the August Release is July 15th. If you want this fix to be included in that release, please merge this PR before that date.

**Comment by carlossanlop** [comment]
Friendly reminder that Monday July 15th is Code Complete day, that's the deadline to get this included in the August Release.

**Pull request closed without merging**